src/components/AttachmentPicker/index.native.js

```diff
@@ -94,43 +124,135 @@ function getDataForUpload(fileData) {
     };
 }
 
-const propTypes = {
-    children: PropTypes.func.isRequired,
-};
-
 /**
  * This component renders a function as a child and
  * returns a "show attachment picker" method that takes
- * a callback. This is the web/mWeb/desktop version since
- * on iOS Safari we must append a hidden input to the DOM
- * and listen to onChange event. When the show method is
- * called an attachment
- *
- * @example
- * <AttachmentPicker>
- * {({openPicker}) => (
- *     <Button
- *         onPress={() => {
- *             openPicker({
- *                 onPicked: (file) => {
- *                     // Display or upload File
- *                 },
- *             });
- *         }}
- *     />
- * )}
- * </AttachmentPicker>
- *
- * @returns {Function}
+ * a callback. This is the ios/android implementation
+ * opening a modal with attachment options
  */
-const AttachmentPicker = ({children}) => children({
-    openPicker: ({onPicked}) => {
-        show((response) => {
-            onPicked(getDataForUpload(response));
+class AttachmentPicker extends Component {
+    constructor(...args) {
+        super(...args);
+
+        this.state = {
+            isVisible: false,
+            result: null,
+            onPicked: () => {},
+        };
+
+        this.menuItemData = [
+            {
+                icon: Camera,
+                text: 'Take Photo',
+                pickAttachment: () => showImagePicker(RNImagePicker.launchCamera),
+            },
+            {
+                icon: Gallery,
+                text: 'Choose from Gallery',
+                pickAttachment: () => showImagePicker(RNImagePicker.launchImageLibrary),
+            },
+            {
+                icon: Paperclip,
+                text: 'Choose Document',
+                pickAttachment: showDocumentPicker,
+            },
+        ];
+
+        this.setResult = this.setResult.bind(this);
+        this.close = this.close.bind(this);
+        this.completeAttachmentSelection = this.completeAttachmentSelection.bind(this);
+    }
+
+    /**
+     * Store the selected attachment mapped to an appropriate file interface
+     *
+     * @param {ImagePickerResponse|DocumentPickerResponse} attachment
+     */
+    setResult(attachment) {
+        if (attachment && !attachment.didCancel && !attachment.error) {
+            const result = getDataForUpload(attachment);
+            this.setState({result});
+        }
+    }
+
+    /**
+     * Triggers the `onPicked` callback with the selected attachment
+     */
+    completeAttachmentSelection() {
+        if (this.state.result) {
+            this.state.onPicked(this.state.result);
+        }
+    }
+
+    /**
+     * Opens the attachment modal
+     *
+     * @param {function} onPicked A callback that will be called with the selected attachment
+     * @throws {Error} when the `onPicked` parameter is invalid
+     */
+    open(onPicked) {
+        if (typeof onPicked !== 'function') {
```

Comment:
Is there a known scenario where we could be passing a non-function parameter to `open`?

Reply:
No. And there shouldn't be
 
This is an insurance that whoever is using the component is notified right away if they haven't set it up correctly similar to prop types warnings, but prop types can only declare `children` is a required function and they can't validate it'll be called correctly

If we don't have an error guard here, you'll this error will just happen later when you try to call `onPicked` after selecting an attachment it's better to shoot right away so this doesn't slip undetected

I'm leaving the comment unresolved so you can verify this addresses your concern

Reply:
I'm resolving this as it seems to be the only thing blocking the PR

Reply:
I get that this might help someone, but if that is its only purpose then it doesn't seem necessary. If someone creates the component elsewhere and sees an `: TypeError: undefined is not a function`, it should be pretty obvious where the issue lies.

Reply:
> I get that this might help someone, but if that is its only purpose then it doesn't seem necessary. If someone creates the component elsewhere and sees an : TypeError: undefined is not a function, it should be pretty obvious where the issue lies.

If this check is removed `undefined is not a function` would only appear after you complete the flow and select a file. 

With or without this clause you will end up with an error if the component is not setup properly, the check would just bring it to your attention sooner

I see this as something that can only help. As an external contributor it was my first time seeing a modal setup in this particular render prop configuration and it was a bit confusing for me. 

Anyway if it doesn't seem necessary to you I'll remove it next week, just let me know 😉

Reply:
I appreciate the reasoning. Including this might cause other to add similar pre-optimisations which we tend to try to avoid. It's not a huge problem, but better to remove it imo.

Reply:
No problem, I'll remove it